src/aws4_request.gleam

```diff
@@ -1,28 +1,116 @@
+//// Sign requests with AWS Signature Version 4 for AWS API.
+
 import gleam/crypto
 import gleam/http
 import gleam/http/request.{type Request, Request}
 import gleam/int
 import gleam/list
 import gleam/bit_array
-import gleam/option
+import gleam/option.{type Option, None, Some}
 import gleam/string
 
-// TODO: document
-// TODO: document params
-// https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
-// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
-pub fn sign(
+/// A custom DateTime type in #((year, month, day), (hour, minute, second)) format.
+pub type DateTime =
+  #(#(Int, Int, Int), #(Int, Int, Int))
+
+/// Signs a Request(BitArray) with AWS Signature Version 4 by adding signed headers,
+/// and returns the signed request.
+///
+/// See: https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html
+///
+///   Parameters:
+/// request: The request to sign of type Request(BitArray).
+/// date_time: The current datetime in #((year, month, day), (hour, minute, second)) format.
+/// access_key_id: The AWS credential access key ID being used to authenticate the request.
+/// secret_access_key: The AWS credential secret access key being used to authenticate the
+/// request.
+/// session_token: (Optional) The security token value if using temporary credentials from AWS STS.
+/// region: The AWS region that the request will be made to.
+/// service: The AWS service name that the request is for.
```

Comment:
I saw that parameters needed to be documented but I was unable to find an example of the correct format in the language documentation or in source code for the libraries I've seen so far.

Originally I put these `///` comments directly above the line of each parameter, but `gleam format` puts them all up here so I went with this, but please link me if there is an example of how to properly do this!